### PR TITLE
Only update cluster-ca-key-generation from CaReconciler

### PR DESCRIPTION
**Pull request by katheris**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

When rolling pods to trust new CA cert signed by new CA key only update the cluster-ca-key-generation
annotation. Do not update cluster-ca-cert-generation or clients-ca-cert-generation.

In PR #10711 I added logic to update the cluster-ca-cert-gen , clients-ca-cert-gen and cluster-ca-key-gen annotations before rolling the pods. However, the pods are being rolled to trust a new CA cert that is signed by a new CA key. They are still presenting certificates that were issued by the old CA cert.

This means that if the reconcile fails after rolling the Kafka pods as part of the CaReconciler rolling update, but before it runs the KafkaReconciler logic, when it comes back in it would incorrectly think that the new CA cert was used everywhere and remove the old one from the ca-cert Secret.

We should only update the cluster-ca-key-gen annotation before doing the rolling update

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Review by ppatierno — APPROVED** [review]

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java [hunk 1]
```diff
@@ -575,6 +575,15 @@ Future<Void> rollingUpdateForNewCaKey() {
...
+                // Check if this pod has been rolled previously
+                int clusterCaKeyGeneration = clusterCa.keyGeneration();
+                int podClusterCaKeyGeneration = Annotations.intAnnotation(pod, Ca.ANNO_STRIMZI_IO_CLUSTER_CA_KEY_GENERATION, clusterCaKeyGeneration);
+                if (clusterCaKeyGeneration == podClusterCaKeyGeneration) {
+                    LOGGER.debugCr(reconciliation, "Not rolling Pod {} since the CA cert key generation is correct.", pod.getMetadata().getName());
+                    return RestartReasons.empty();
+                }
+            }
             LOGGER.debugCr(reconciliation, "Rolling Pod {} due to {}", pod.getMetadata().getName(), podRollReasons.getReasons());
             return podRollReasons;
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java:588]: It is better to wrap this into `else` branch to have return validation.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java [hunk 2]
```diff
@@ -575,6 +575,15 @@ Future<Void> rollingUpdateForNewCaKey() {
                 false,
                 eventPublisher
         ).rollingRestart(pod -> {
+            if (podRollReasons.getReasons().size() == 1 && podRollReasons.contains(RestartReason.CLUSTER_CA_CERT_KEY_REPLACED)) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java:578]: Why is this needed? Why do we even need the rolling update here for Clients CA key replacement?
* It is not used internally, so we do not need to establish trust
* It should be enough to roll this in the regular rolling update where it would anyway roll to change the annotation, or?

**Comment by katheris** [comment]
I'm investigating the failure, I've worked out what's causing it, just trying to determine how best to address it

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java [hunk 3]
```diff
@@ -579,8 +578,21 @@ Future<Void> rollingUpdateForNewCaKey() {
...
+                    LOGGER.debugCr(reconciliation, "Not rolling Pod {} since the CA cert key generation is correct.", pod.getMetadata().getName());
+                    return RestartReasons.empty();
+                } else {
+                    LOGGER.debugCr(reconciliation, "Rolling Pod {} due to {}", pod.getMetadata().getName(), podRollReasons.getReasons());
+                    return podRollReasons;
+                }
+            } else {
+                LOGGER.debugCr(reconciliation, "Rolling Pod {} due to {}", pod.getMetadata().getName(), podRollReasons.getReasons());
+                return podRollReasons;
+            }
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java:595]: So, does this need to be updated to handle both reasons = both annotations? Otherwise, why are we patching the clients CA generation in the PodSet?

**Comment by katheris** [comment]
@scholzj @ppatierno I've pushed some updates to only roll when the cluster CA key is replaced, not the clients CA key (as we can leave that to the KafkaRoller) and only use the annotations, not the `keyReplaced()` method. I haven't done any manual testing yet, so take a look if you want but if you'd prefer to wait til I've done some manual testing for the clients CA key I'll update here when I've done that.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/RestartReason.java [hunk 4]
```diff
@@ -28,11 +28,6 @@ public enum RestartReason {
      */
     CA_CERT_RENEWED("CA certificate renewed"),
 
-    /**
-     * Clients CA private key was replaced
-     */
-    CLIENT_CA_CERT_KEY_REPLACED("Trust new clients CA certificate signed by new key"),
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/RestartReason.java:34]: I guess this should be removed from docs as well (search for `ClientCaCertKeyReplaced`).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java [hunk 5]
```diff
@@ -376,31 +376,19 @@ Future<Void> reconcileClusterOperatorSecret(Clock clock) {
...
-     * due to a new CA key. It is not necessary when the CA certificate is replace while retaining the existing key.
+     * certificates get removed from their truststores. Note this is only necessary when the Cluster CA certificate has changed
+     * due to a new CA key. It is not necessary when the CA certificate is replaced while retaining the existing key.
      */
-    Future<Void> rollingUpdateForNewCaKey() {
-        RestartReasons podRollReasons = RestartReasons.empty();
-
-        // cluster CA needs to be fully trusted
-        // it is coming from a cluster CA key replacement which didn't end successfully (i.e. CO stopped) and we need to continue from there
-        if (clusterCa.keyReplaced() || isClusterCaNeedFullTrust) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java:387]: I hope we are 100% sure the `clusterCa.keyReplaced()` is not needed :-o

**Comment by scholzj** [comment]
@katheris Looks like the tests related to CA renewal failed: `SecurityST.testAutoReplaceClientsCaKeysTriggeredByAnno`

**Comment by katheris** [comment]
Since the change this is based on had to be reverted I'm closing this PR and will open a new one with the changes all in a single PR
